contracts/VetoERC20Voting.sol

```diff
@@ -54,16 +77,107 @@ contract VetoERC20Voting is IVetoERC20Voting, TransactionHasher, Initializable {
+
         // User has voted
         userHasVoted[msg.sender][_transactionHash] = true;
 
-        emit VetoVoteCast(msg.sender, _transactionHash, vetoVotes);
+        emit VetoVoteCast(msg.sender, _transactionHash, vetoVotes, _freeze);
+    }
+
+    /// @notice Allows a user to cast a freeze vote if there is an active freeze proposal
+    /// @notice If there isn't an active freeze proposal, it is created and the user's votes are cast
+    function castFreezeVote() public {
+        uint256 userVotes;
+
+        if (
+            block.number >
+            freezeProposalCreatedBlock + freezeProposalBlockDuration
+        ) {
+            // Create freeze proposal, set total votes to msg.sender's vote count
+            freezeProposalCreatedBlock = block.number;
+
+            userVotes = votesToken.getPastVotes(
+                msg.sender,
+                freezeProposalCreatedBlock - 1
+            );
+            require(userVotes > 0, "User has no votes");
+
+            freezeProposalVoteCount = userVotes;
+
+            emit FreezeProposalCreated(msg.sender);
+        } else {
+            // There is an existing freeze proposal, count user's votes
+            require(
+                !userHasFreezeVoted[msg.sender][freezeProposalCreatedBlock],
+                "User has already voted"
+            );
+
+            userVotes = votesToken.getPastVotes(
+                msg.sender,
+                freezeProposalCreatedBlock - 1
+            );
+            require(userVotes > 0, "User has no votes");
+
+            freezeProposalVoteCount += userVotes;
+        }
+
+        userHasFreezeVoted[msg.sender][freezeProposalCreatedBlock] = true;
+
+        emit FreezeVoteCast(msg.sender, userVotes);
+    }
+
+    /// @notice Unfreezes the DAO, only callable by the owner
+    function defrost() public onlyOwner {
+        require(isFrozen(), "DAO is not already frozen");
+        freezeProposalCreatedBlock = 0;
+        freezeProposalVoteCount = 0;
+    }
+
+    /// @notice Updates the veto votes threshold, only callable by the owner
+    /// @param _vetoVotesThreshold The number of votes required to veto a transaction
+    function updateVetoVotesThreshold(uint256 _vetoVotesThreshold)
+        external
+        onlyOwner
+    {
+        vetoVotesThreshold = _vetoVotesThreshold;
     }
 
-    /// @notice Returns whether the specified functions has been vetoed
+    /// @notice Updates the freeze votes threshold, only callable by the owner
+    /// @param _freezeVotesThreshold Number of freeze votes required to activate a freeze
+    function updateFreezeVotesThreshold(uint256 _freezeVotesThreshold)
+        external
+        onlyOwner
+    {
+        freezeVotesThreshold = _freezeVotesThreshold;
+    }
+
+    /// @notice Updates the freeze proposal blocks duration, only callable by the owner
+    /// @param _freezeProposalBlockDuration The number of blocks a freeze proposal has to succeed
+    function updateFreezeProposalBlockDuration(
```

Comment:
having this method is also kind of strange to me, it effectively means if a vote is nearing completion, and the DAO doesn't get the option it wants, they can tack on more time?

Reply:
Yep, I also don't see this as an issue. This would be the parent DAO changing how long they have to vote on whether the child DAO gets frozen. I'm viewing the parent DAO as being somewhat "all-powerful", which may be an incorrect assumption, but maybe we need to discuss that power balance further.

Reply:
yeah, agreed.  I find it kind of weird, but DAOs should be able to elect not to have that power, or relinquish it if they would like to

Reply:
@herbig what is the core reason as to why you view the scenario you suggested, as a bad thing?

Reply:
Eh, never mind, this was discussed in another comment thread.